File: app/src/java/java/ru/esm/tspiot/dkkt/service/ControlledEncryptionResult.java
package ru.esm.tspiot.dkkt.service;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Objects;

public class ControlledEncryptionResult implements Parcelable {
    private final byte[] mData;

    public ControlledEncryptionResult(byte[] mData) {
        this.mData = mData;
    }

    public byte[] getData() {
        return mData;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        ControlledEncryptionResult that = (ControlledEncryptionResult) o;
        return Objects.deepEquals(mData, that.mData);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(mData);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull @NotNull Parcel dest, int flags) {
        dest.writeByteArray(mData);
    }

    public static final Creator<ControlledEncryptionResult> CREATOR = new Creator<>() {
        @Override
        public ControlledEncryptionResult createFromParcel(Parcel in) {
            return new ControlledEncryptionResult(in.createByteArray());
        }

        @Override
        public ControlledEncryptionResult[] newArray(int size) {
            return new ControlledEncryptionResult[size];
        }
    };
}

File: app/src/kotlin/kotlin/ru/esm/tspiot/dkkt/service/ControlledEncryptionResult.kt
package ru.esm.tspiot.dkkt.service

import android.os.Parcel
import android.os.Parcelable

data class ControlledEncryptionResult(
    val encryptedData: ByteArray,
    val statusCode: Int,
    val statusMessage: String
) : Parcelable {
    constructor(parcel: Parcel) : this(
        parcel.createByteArray() ?: byteArrayOf(),
        parcel.readInt(),
        parcel.readString() ?: ""
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeByteArray(encryptedData)
        parcel.writeInt(statusCode)
        parcel.writeString(statusMessage)
    }

    override fun describeContents(): Int = 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ControlledEncryptionResult

        if (!encryptedData.contentEquals(other.encryptedData)) return false
        if (statusCode != other.statusCode) return false
        if (statusMessage != other.statusMessage) return false

        return true
    }

    override fun hashCode(): Int {
        var result = encryptedData.contentHashCode()
        result = 31 * result + statusCode
        result = 31 * result + statusMessage.hashCode()
        return result
    }

    companion object CREATOR : Parcelable.Creator<ControlledEncryptionResult> {
        override fun createFromParcel(parcel: Parcel): ControlledEncryptionResult =
            ControlledEncryptionResult(parcel)

        override fun newArray(size: Int): Array<ControlledEncryptionResult?> =
            arrayOfNulls(size)
    }
}